composer/utils/checkpoint.py

```diff
@@ -248,88 +235,78 @@ def download_checkpoint(
-        composer_states_filepath = os.path.join(extracted_checkpoint_folder, _COMPOSER_STATES_FILENAME)
+        # DeepSpeed checkpoints OR checkpoints of all formats with compression
+        extracted_tar_checkpoint_folder = os.path.join(node_checkpoint_folder, 'checkpoint')
+        composer_states_filepath = os.path.join(extracted_tar_checkpoint_folder, _COMPOSER_STATES_FILENAME)
+    elif fsdp_sharded_state_dict_enabled:
+        # FSDP sharded state dict has a Composer state dict per rank
+        composer_states_filepath = rank_n_checkpoint_filepath
     else:
-        # it's not an archive; it's just the composer state dict
-        # and only rank zero has this file unless fsdp_sharded_state_dict_enabled then
-        # every rank has it's own file.
-        extracted_checkpoint_folder = None
-        composer_states_filepath = (rank_n_checkpoint_filepath
-                                    if fsdp_sharded_state_dict_enabled else rank_zero_checkpoint_filepath)
+        # Vanilla Composer has a single Composer state dict on local rank zero
+        composer_states_filepath = os.path.join(node_checkpoint_folder, 'rank0_checkpoint')
 
-    try:
-        if ((fsdp_sharded_state_dict_enabled and dist.get_global_rank() == 0) or
-            (not fsdp_sharded_state_dict_enabled and dist.get_local_rank() == 0)):
-            # every NODE needs the GLOBAL rank zero checkpoint unless fsdp_sharded_state_dict_enabled.
-            path = _format_path_with_rank_zero(path)
-            get_file(destination=rank_zero_checkpoint_filepath,
-                     path=path,
-                     object_store=object_store,
-                     progress_bar=progress_bar)
-            if extracted_checkpoint_folder is not None:
-                try:
-                    with tarfile.open(rank_zero_checkpoint_filepath) as tarball:
-                        tarball.extractall(extracted_checkpoint_folder)
-                except FileNotFoundError:
-                    # Not re-raising the file-not-found error as that is irrelevant;
-                    # the underlying issue is that the checkpoint file does not exist on the disk
-                    # or could not be downloaded
-                    raise RuntimeError(f'Checkpoint {path} does not exist')
-
-        if rank_zero_checkpoint_filepath != rank_n_checkpoint_filepath:
-            # every RANK needs ITS OWN checkpoint.
-            # But, the  global rank zero is a special case -- these files are the same!
-            assert dist.get_global_rank() != 0, 'invariant violation'
+    local_path = _format_path_with_current_rank(path)
+    local_rank_zero_path = _get_local_rank_zero_path(local_path)
+    logging.log(logging.WARN, f'local_path={local_path}, local_rank_zero_path={local_rank_zero_path}')
 
+    try:
+        # Download on local rank 0 or if path is different from local rank 0
+        if dist.get_local_rank() == 0 or local_path != local_rank_zero_path:
```

Comment:
shouldn't only deepspeed and sharded checkpoints try get file on all ranks?

Reply:
If you format your path as `...rank{rank}.pt` and it's different from rank 0, we try to download anyways. A user could be using a weird checkpoint which is different per rank for some reason, and if they pass that in, we try it. We don't raise an error on other ranks if the file doesn't exist (as we expect it to not exist)